feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt

```diff
@@ -467,22 +464,26 @@ class BillingWrapper(
         }
     }
 
-    internal fun getPurchaseType(purchaseToken: String): ProductType {
+    internal fun getPurchaseType(purchaseToken: String, listener: (ProductType) -> Unit) {
         billingClient?.let { client ->
-            val querySubsResult = client.queryPurchases(SkuType.SUBS)
-            val subsResponseOK = querySubsResult.responseCode == BillingClient.BillingResponseCode.OK
-            val subFound = querySubsResult.purchasesList?.any { it.purchaseToken == purchaseToken } ?: false
-            if (subsResponseOK && subFound) {
-                return@getPurchaseType ProductType.SUBS
-            }
-            val queryInAppsResult = client.queryPurchases(SkuType.INAPP)
-            val inAppsResponseIsOK = queryInAppsResult.responseCode == BillingClient.BillingResponseCode.OK
-            val inAppFound = queryInAppsResult.purchasesList?.any { it.purchaseToken == purchaseToken } ?: false
-            if (inAppsResponseIsOK && inAppFound) {
-                return@getPurchaseType ProductType.INAPP
+            client.queryPurchasesAsync(SkuType.SUBS) { querySubsResult, subsPurchasesList ->
+                val subsResponseOK = querySubsResult.responseCode == BillingClient.BillingResponseCode.OK
+                val subFound = subsPurchasesList.any { it.purchaseToken == purchaseToken }
+                if (subsResponseOK && subFound) {
```

Comment:
hmmm i believe most of the responses we'd wanna retry on involve the service disconnecting, which i THINK would trigger `onBillingServiceDisconnected` first and we'd retry connection with andy's [latest changes](https://github.com/RevenueCat/purchases-android/pull/558/files)...i'd have to test that `onBillingServiceDisconnected` gets called first, though, so that this automatically gets retried.

@vegaro what do you think?

for reference, here are the potential response codes:
```        int SERVICE_TIMEOUT = -3;
        int FEATURE_NOT_SUPPORTED = -2;
        int SERVICE_DISCONNECTED = -1;
        int OK = 0;
        int USER_CANCELED = 1;
        int SERVICE_UNAVAILABLE = 2;
        int BILLING_UNAVAILABLE = 3;
        int ITEM_UNAVAILABLE = 4;
        int DEVELOPER_ERROR = 5;
        int ERROR = 6;
        int ITEM_ALREADY_OWNED = 7;
        int ITEM_NOT_OWNED = 8;```

Reply:
I agree that we would want to retry the responses that involve the service disconnecting, but I am not sure either `onBillingServiceDisconnected` will be triggered. And in any case, with our current code we would just try to reconnect and not try `client.queryPurchasesAsync` again. I am also not sure if `client.queryPurchasesAsync` can even return response code is `SERVICE_DISCONNECTED`. In that case we should probably retry the `client.queryPurchasesAsync`. Andy's changes only try to reconnect if the setup fails.

Maybe we should add checks after all calls to the billing client to check if the response code is `SERVICE_DISCONNECTED` ? Definitely not something I think should be tackled in this PR

Reply:
> Maybe we should add checks after all calls to the billing client to check if the response code is SERVICE_DISCONNECTED ?

Well, I think it wouldn't be bad but I would prefer to know if this is actually happening in the wild before doing it. Maybe if we had more information about the type of errors users get (RevenueCatlytics?) we could decide whether this is necessary.

In any case, I agree it's not something to be done in this PR and it's fine as it is for now.

Reply:
RevenueCatlytics++, that's exactly the use case I most want it for...logging errors that come from the store.